clusters/build-clusters/01_cluster/openshift-ingress-operator/REAME.md

```diff
@@ -0,0 +1,59 @@
+# openshift-ingress-operator
+
+## CA Certificate for app routes
+
+Openshift 4.2 has doc on [this topic](https://docs.openshift.com/container-platform/4.2/authentication/certificates/replacing-default-ingress-certificate.html).
+
+### Manual steps
+
+#### AWS IAM configuration
+
+[set up aws IAM user](https://github.com/Neilpang/acme.sh/wiki/How-to-use-Amazon-Route53-API): user `openshift-ci-robot` in group `acme` which
+    has the policy `acme`.
+
+#### Generate the certificate
+
+Using [acme.sh](https://github.com/Neilpang/acme.sh)
+
+```bash
+$ podman run -it --rm neilpang/acme.sh /bin/ash
+### BitWarden: aws_ci_infra_openshift-ci-robot
+# export  AWS_ACCESS_KEY_ID=xxx
+# export  AWS_SECRET_ACCESS_KEY=yyy
+# acme.sh --issue --dns dns_aws -d api.build01.ci.devcluster.openshift.com -d '*.apps.build01.ci.devcluster.openshift.com'
+...
+[Sun Nov 24 18:24:26 UTC 2019] Your cert is in  /acme.sh/*.apps.build01.ci.devcluster.openshift.com/*.apps.build01.ci.devcluster.openshift.com.cer 
+[Sun Nov 24 18:24:26 UTC 2019] Your cert key is in  /acme.sh/*.apps.build01.ci.devcluster.openshift.com/*.apps.build01.ci.devcluster.openshift.com.key 
+[Sun Nov 24 18:24:26 UTC 2019] The intermediate CA cert is in  /acme.sh/*.apps.build01.ci.devcluster.openshift.com/ca.cer 
+[Sun Nov 24 18:24:26 UTC 2019] And the full chain certs is there:  /acme.sh/*.apps.build01.ci.devcluster.openshift.com/fullchain.cer
+...
+```
+
+> $ oc create secret tls app-cert --cert=path/to/fullchain.cer --key=/path/to/the.key  -n openshift-ingress --dry-run -o yaml | oc apply -f -
+
+
+> oc patch ingresscontroller.operator default \
+     --type=merge -p \
+     '{"spec":{"defaultCertificate": {"name": "app-cert"}}}' \
+     -n openshift-ingress-operator
+
+Verify if it works:
+
+```
+$ curl --insecure -v https://default-route-openshift-image-registry.apps.build01.ci.devcluster.openshift.com 2>&1 | awk 'BEGIN { cert=0 } /^\* Server certificate:/ { cert=1 } /^\*/ { if (cert) print }'
+* Server certificate:
+*  subject: CN=*.apps.build01.ci.devcluster.openshift.com
+*  start date: Nov 24 17:24:24 2019 GMT
+*  expire date: Feb 22 17:24:24 2020 GMT
+*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
+*  SSL certificate verify ok.
+* Mark bundle as not supporting multiuse
+* Added cookie 34727b82525eb26a530629c5bf0ec2f2="3a8eed8be69a9366179c6e84d1dd4f30" for domain default-route-openshift-image-registry.apps.build01.ci.devcluster.openshift.com, path /, expire 0
+* Connection #0 to host default-route-openshift-image-registry.apps.build01.ci.devcluster.openshift.com left intact
+
+```
+
+TODO: solution to auto-renewing the certs.
+
+* The certs are valid for 90 days.
+* After renewal, the certs are changed. We need to apply the new certs to the secret above. Re-issuing might be simpler in our case.
```

Comment:
@stevekuznetsov 

How about running the above manual steps in a pod weekly?

Reply:
What's the difference between re-issuing and applying the new certs?

Reply:
`acme` is the client side script:

* acme script tells letsencrypt that some domain need a certificate
* letsencrypt issues a challenge and acme script adds it (as txt record of the aws hosted domain) and waits for letsencrypt to verify.
* when letsencrypt is satisfied with txt record for the domain, acme script can download cert files and clean the txt records.

The above is the procedure of issuing ... 

1. Reissuing: run the above procedure again without anything cached. In this case, my understanding is that we will get a _new_ set of certs (need test to make sure). Would it make the previous certs expired? I think so.
2. acme can also do `renewal` of certs (with the current certs cached), but the renewal does not run if the cached certs is still valid for a long time (i think it has to be the last 30 days to renews, but I need to wait till that day to test)

In the acmd container (with the docker image above) ... there is a cron job set up already to do the renewal, but we are not going to run the container as a server, instead it is a periodic job.

Why I do not want to cache the current certs? it is a folder holding a bunch of files, and it is hard to mount all of them from secret when periodic runs.
Re-Issuing only when necessary (instead of reissuing blindly): expiring in 30 days so that we do not need update openshift secret so often and we still have enough time to fix if something does not work